### PR TITLE
Tibor/merge all fld char support

### PR DESCRIPTION
**Pull request by tiberlas**
## Description

- added support for complex field char
This solution check for the relationship between the current run and posible fldChars in the paragraph of the run.

Closes: [#4415](https://github.com/openlawlibrary/platform/issues/4415)

## Code review checklist

- [ ] Private platform tests at `core/tests/test_python-docx` are updated and passing
- [ ] If this change is going to be deployed on Cloudsmith after merge, python-docx version number should be increased 

[commit messages]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by tiberlas** [comment]
I have created a test file that uses cProfile to measure how long is needed to read 18000 paragraphs, and these are the result:
- currently the code on 'merge-all' branch has a cumulative time of 4.63 seconds
- this code on the PR has a cumulative time of 6 seconds

The major down side of this solution is the use of lxml when we have a cache miss.
Cache miss occur in two ways:
1. when we're reading the first run of the paragraph
2. when we're reading a run in the middle of a new paragraph, e.g. `doc.paragraphs[X].runs[Y].text`

Possible improvement is to add a flag to `doc.paragraphs[X].text` and then we can resolve the first major downside, because we know that we are reading the hole paragraph from the begging. But this will result in a messier code.

File: docx/oxml/text/paragraph.py [hunk 1]
```diff
@@ -128,9 +128,39 @@ def iter_r_lst_recursive(self):
...
+                            if t == 'begin':
+                                ignoreRun += 1
+                            elif t == 'separate':
+                                ignoreRun -= 1
+                                # should ignore the next `end`, because text
+                                # is shown from `separate` to `end`
+                                hasSeparate = True
+                                # We know this `fldCharType == 'end'`, so we check
+                                # if this `fldChar` has 'separate' in it.
+                            elif hasSeparate == False:
```
Comment (balsa-sarenac) [inline docx/oxml/text/paragraph.py]: In python it's better to use `is False` than `== False`. Check this out: https://stackoverflow.com/a/27276841

File: docx/oxml/text/paragraph.py [hunk 2]
```diff
@@ -128,9 +128,39 @@ def iter_r_lst_recursive(self):
...
+                                ignoreRun -= 1
+                            else:
+                                hasSeparate = False
+                    # removes ``<w:fldChar>`` form run because the hidden elements are removed and this tag is obsolete
+                    for child_r in child:
+                        if child_r.tag == qn('w:fldChar'):
+                            child.remove(child_r)
+                    # if run has no visible element that do not return it
```
Comment (balsa-sarenac) [inline docx/oxml/text/paragraph.py]: Maybe reword it a bit like "yields runs that have at least one visible element"

File: docx/oxml/text/paragraph.py [hunk 3]
```diff
@@ -128,9 +128,39 @@ def iter_r_lst_recursive(self):
...
+                                # We know this `fldCharType == 'end'`, so we check
+                                # if this `fldChar` has 'separate' in it.
+                            elif hasSeparate == False:
+                                ignoreRun -= 1
+                            else:
+                                hasSeparate = False
+                    # removes ``<w:fldChar>`` form run because the hidden elements are removed and this tag is obsolete
+                    for child_r in child:
+                        if child_r.tag == qn('w:fldChar'):
+                            child.remove(child_r)
```
Comment (balsa-sarenac) [inline docx/oxml/text/paragraph.py:160]: Why is this needed? Do `w:fldChar` elements have any content that can be previewed? How are tests behaving with/without this?

File: docx/text/paragraph.py [hunk 4]
```diff
@@ -427,8 +427,9 @@ def set_li_lvl(self, styles, prev, ilvl):
     @property
     @cache
     def run_text(self):
-        if self.runs:
-            return ''.join(r.text for r in self.runs)
+        runs = self.runs
+        if runs:
+            return ''.join(r.text for r in runs)
```
Comment (balsa-sarenac) [inline docx/text/paragraph.py:432]: I assume you made this change for performance reasons?

File: docx/oxml/text/paragraph.py (hunk 3 above)
Comment (tiberlas) [reply docx/oxml/text/paragraph.py:160]: We need to remove `w:fldChar` because of the cache. So if we are reading the same paragraph twice that we do not have to do the logic again.

File: docx/text/paragraph.py (hunk 4 above)
Comment (tiberlas) [reply docx/text/paragraph.py:432]: That is correct.

**Review by balsa-sarenac — APPROVED** [review]
